### PR TITLE
Better handling of HTTP error 503

### PR DESCRIPTION
**Pull request by dmitry-yudakov**
Redirecting to maintenance page is pretty bad user experience - especially after momentum has been load and was working successfully. You cannot simply reload the page because of changed url and the existing state potentially working gets lost.

That's why I made a change - if the first request (`config/ui-client`) fails with 503 or something else, then we show Maintenance or SmtWentWrong error without changing url.

Further API requests with 503 are retried up to 3 times with incremental delay (1 sec, 2 sec, 4 sec) - configurable. If it still fails after retry, toast notification is shown.

It's implemented without using external libraries as they didn't play too well with our existing logic.

### CODE REVIEW

File: packages/app/src/scenes/App.tsx [hunk 1]
```diff
@@ -34,10 +34,20 @@ const App: FC = () => {
   const {t} = useTranslation();
 
   useEffect(() => {
+    setAxiosRetry({
```
Comment (PAKovalev) [inline packages/app/src/scenes/App.tsx]: Lets create a hook instead of useEffect

File: packages/app/src/api/request/request.ts [hunk 2]
```diff
@@ -53,8 +53,48 @@ const getAccessToken = (): string => {
...
-export const setApiResponseHandlers = ({onResponse = responseHandler, onError = errorHandler}) => {
-  request.interceptors.response.use(onResponse, onError);
+type AxiosRequestConfigWithRetryType = AxiosRequestConfig & {momentumRetryCount?: number};
+
+export const setApiResponseHandlers = ({
+  onResponse = responseHandler,
+  onError = errorHandler,
+  maxRetries = 3,
+  retryDelayBase = 1000,
+  retryCodes = [503]
```
Comment (PAKovalev) [inline packages/app/src/api/request/request.ts]: Use existing constants.

File: packages/app/src/api/request/request.ts [hunk 3]
```diff
@@ -53,8 +53,48 @@ const getAccessToken = (): string => {
...
 request.interceptors.request.use(requestHandler, errorHandler);
 
-export const setApiResponseHandlers = ({onResponse = responseHandler, onError = errorHandler}) => {
-  request.interceptors.response.use(onResponse, onError);
+type AxiosRequestConfigWithRetryType = AxiosRequestConfig & {momentumRetryCount?: number};
+
+export const setApiResponseHandlers = ({
+  onResponse = responseHandler,
+  onError = errorHandler,
+  maxRetries = 3,
```
Comment (PAKovalev) [inline packages/app/src/api/request/request.ts]: Let's move this to the env file.

File: packages/app/src/scenes/App.tsx [hunk 4]
```diff
@@ -33,24 +33,39 @@ import 'react-toastify/dist/ReactToastify.css';
...
+    isConfigReady,
+    isError: isErrorLoadingConfig,
+    errorCode: configLoadingErrorCode
+  } = configStore;
 
   const {pathname} = useLocation<{pathname: string}>();
   const history = useHistory();
   const {t} = useTranslation();
 
   useEffect(() => {
```
Comment (PAKovalev) [inline packages/app/src/scenes/App.tsx:46]: Lets create a hook under `shared/hooks`.

**Review by PAKovalev — APPROVED** [review]

File: packages/app/src/api/request/request.ts [hunk 5]
```diff
@@ -34,7 +34,7 @@ const responseHandler = (response: AxiosResponse) => {
 };
 
 const errorHandler = (error: AxiosError) => {
-  throw error;
+  return Promise.reject(error) as unknown;
```
Comment (Daboo1998) [inline packages/app/src/api/request/request.ts]: why not change to async and then throw error?

File: packages/app/src/stores/ConfigStore/ConfigStore.ts [hunk 6]
```diff
@@ -25,7 +25,7 @@ const ConfigStore = types
 
         Object.entries(appVariablesOverrides).forEach((entry) => {
           const [key, value] = entry;
-          appVariables[key as keyof AppConfigInterface] = value as any;
+          appVariables[key as keyof AppConfigInterface] = value as string;
```
Comment (Daboo1998) [inline packages/app/src/stores/ConfigStore/ConfigStore.ts:28]: actually this is wrong. It should be 
```suggestion
          appVariables[key as keyof AppConfigExtendedInterface] = value as any;
```

as appVariables are of type `AppConfigExtendedInterface`

**Review by Daboo1998 — CHANGES_REQUESTED** [review]

File: packages/app/src/stores/ConfigStore/ConfigStore.ts (hunk 6 above)
Comment (dmitry-yudakov) [reply packages/app/src/stores/ConfigStore/ConfigStore.ts:28]: It doesn't really matter - it's just casting of stuff and doesn't affect anything. And `any` will give us warning.

File: packages/app/src/scenes/App.tsx (hunk 4 above)
Comment (dmitry-yudakov) [reply packages/app/src/scenes/App.tsx:46]: I'll leave it as is for the moment. Lets discuss when you're back. I don't think it's good to hide all the logic of this effect in a hook, yet if it's just to spare useEffect+setApiResponseHandlers it's meaningless and only gives you pain of setting up the types and figuring out how to solve the issue of the inner deps list of the same useEffect.

File: packages/app/src/stores/ConfigStore/ConfigStore.ts (hunk 6 above)
Comment (dmitry-yudakov) [reply packages/app/src/stores/ConfigStore/ConfigStore.ts:28]: And it gives error `Type 'any' is not assignable to type 'never'.` if AppConfigExtendedInterface is used, whatever the hell it means. Lets just leave it as is ;)

File: packages/app/src/api/request/request.ts (hunk 5 above)
Comment (dmitry-yudakov) [reply packages/app/src/api/request/request.ts]: It was needed for the first implementation, but now I just restored it to the original `throw error`.

File: packages/app/src/api/request/request.ts (hunk 3 above)
Comment (dmitry-yudakov) [reply packages/app/src/api/request/request.ts]: Similar parameters like response timeout are in this file, so I put the new ones next to them.

**Review by Daboo1998 — APPROVED** [review]

**Pull request merged**